### ISSUE TITLE
How promise APIs would look like in core

### ISSUE THREAD

**Issue by balupton** [nodejs/promises#16]
Proposing this thread to consolidate the various discussions and proposals around "How promise APIs would look like in core." as the original PR covers more than just that sole issue, making it hard to keep track of what is actually being proposed. Feel free to edit.

Proposals for current APIs:

- `require('fs').fs.readFilePromise` (current PR proposal) <br/> https://github.com/nodejs/node/pull/5020#issuecomment-183980246
- `require('fs/promise').readFile` vs `require('fs').promised.readFile` (future PR proposal) <br/> https://github.com/nodejs/node/pull/5020#issuecomment-184016883

Proposals for new APIs:

- `util.promised`: https://github.com/nodejs/promises/issues/5
- `util.toCallback`: https://github.com/nodejs/promises/issues/6

**Comment by chrisdickinson** [comment]
Thanks for the issue. I will keep this updated with the current state of the proposal as it evolves.

**Comment by balupton** [comment]
Cool :-)

Some further suggestions:

Perhaps the top post of this thread should go to a wiki page, so we have change history tracked. Not sure if necessary.

The original post of https://github.com/nodejs/node/pull/5020 should probably point to this repo now, especially as many of the links in the original post no longer take me anywhere due to github thread clipping. E.g. clicking the link in "Why use the promisify approach? Response summed up here." doesn't do anything for me. I'm not sure if the quote `I'm going to start cataloguing common questions tonight, and I'll include them under this notice so that you can quickly look to see if there's an answer to your issue without having to search through the entire thread.` is referring to something upcoming, or referring to this repo, or referring to this blog post.

The medium post https://medium.com/@isntitvacant/adding-promise-support-to-core-a4ea895ccbda mentions to PR as the best place to participate, is that still the case, or will it now be this repo?

A new thread to discuss error APIs, that this thread can just link to. Not sure if detaching the two is a good idea or not, but seems the medium post goes into a lot of extra detail about error APIs that we may not want to include here.

**Comment by balupton** [comment]
It would also be good to have links to the why behind the proposals, like:

- Why `require('fs').readFilePromise` + `require('fs').promised.readFile` now then `require('fs/promise').readFile` later? Why not just one? Why not just `require('fs/promise')` now?
- Why was this decision made: `Promises will be added using a promisify wrapper, without affecting the original function. They should not interfere with callback APIs at all.`

The original top post contained some links to discussions, but now it is more a summary now. The medium post goes into plenty more detail, but I can see it becoming out of date once again.

Perhaps consolidating the original post of the PR, the medium post, and the top post here, into one wiki page on this repo, that the WG can keep up to date, and link to the reasonings, then use these individual repo issue threads to facilitate discussions around the particular issues.

**Comment by chrisdickinson** [comment]
@balupton: Part of the issue with those links into the GitHub issue is that they work, but they take a _really long time to load_ — GitHub appears to load the initial comments quickly, but then pauses on an extra XHR before jumping to the anchor, and that appears to take the bulk of the time.

**Comment by benjamingr** [comment]
I'm +1 on `require("fs/promise")`. What are the technical limitations for that?

**Comment by chrisdickinson** [comment]
@benjamingr: The technical requirements should be pretty straightforward given the way that the promise API is exposed now, _but_ there's significant work to be done in bringing a conversation about such an approach to consensus. I believe this would best be accomplished by a follow-on PR once nodejs/node#5020 is merged (assuming it will be!), to be merged before unflagging, which is the current plan of action.

**Comment by balupton** [comment]
@chrisdickinson Could doing multiple PRs mean that each merged PR's API is locked in? That seems to be the risk here.

**Comment by chrisdickinson** [comment]
@balupton I do not think so — we are working behind a flag, and the next step after the initial PR is merged is to build up a "go / no go" checklist for unflagging. We can (if we so choose) delay unflagging on getting a satisfactory answer to the `require('fs/promise')` discussion. The initial PR gets us to the point that we can work behind a flag.

**Comment by benjamingr** [comment]
+1 for `require("fs/promise")` vs `require("fs").promised` blocking unflagging but nor PR.

+1 for error object/`.recover` not blocking the PR.

**Comment by phpnode** [comment]
Bikeshed: I firmly believe it's safer and more sensible to call this `require('fs/2')` instead of `require('fs/promise')`. This gives us a safe path to introducing changes which would otherwise cause compatibility problems in future. For instance if whatwg streams were accepted in future, they could go into `require('stream/2')`. This gives node the ability to evolve its API over time without throwing existing code under the bus.

**Comment by chrisdickinson** [comment]
@phpnode: Noted, but this can be discussed on the subsequent PR. It's not likely that a `/2` moniker will be uncontentious, since that naming scheme implies that callbacks are an older, and thus invalid, pattern — which is not the case.

**Comment by phpnode** [comment]
I agree that it might be contentious but the reason is social not technical, whereas the benefits of using `/2` is technical - it lets us evolve the API. If some unknown new language feature comes along in the next few years (maybe Observables) and we want to adopt that in node, what is our path, especially when there is `fs` and `fs/promise`? This is obvious when modules are explicitly versioned - `fs/3`. 

The counter point is that explicitly requiring `fs/promise` makes promises feel second class, and it is non-obvious to newcomers why they'd need to include that when they're thinking in terms of `async` and `await`, whereas saying "use the latest version of the fs module" is more straightforward.

I don't think it implies that callbacks are invalid, they're not being deprecated, but they are older.

**Comment by benjamingr** [comment]
@phpnode I think that this attitude where callbacks are "older" and promises are "the next API which people should default to using" will not ever get promises into node core.

Callbacks were there first so suggesting `"fs/callbacks"` and `"fs/promises"` is not an option. I'm personally completely fine with `"fs/promise"`

**Comment by phpnode** [comment]
> I think that this attitude where callbacks are "older" and promises are "the next API which people should default to using" will not ever get promises into node core.

Agree, I'm not suggesting that, and definitely not suggesting `fs/callbacks`. The main reason I have a slight problem with `fs/promise` is the future upgrades issue, I don't personally look forward to a world where we have `fs`, `fs/promise` and `fs/observable` for instance. So part of this discussion should probably focus on how those future similar types of changes would be accommodated.

**Comment by benjamingr** [comment]
@phpnode observables would not replace promises but would rather work in areas Node works with streams and event emitters. No one is suggesting replacing all callbacks with observables. 

Observables and async iterators are the problems of the streams working group - not us :)

**Comment by phpnode** [comment]
@benjamingr I'm not talking about observables specifically, rather how this kind of disruptive change is accommodated in future - in this case we're talking about promises, maybe next time we'll be talking about something equally disruptive. Observables is just an example.

**Comment by balupton** [comment]
If compat ever breaks, which is too early to tell, if could always be fs/promises/2 — fragmenting entire APIs under a single versioning scheme this early seems weird. What happens if the callback APIs change, would that be fs/3 which also includes promises but doesn't… doesn't really make sense.

**Comment by balupton** [comment]
Regarding handling operational errors (currently planned via .catch but also some suggesting for throwing for try...catch) vs programmer errors (currently implemented as throwing errors). Why not have programming errors be of their own error type (say EINVALID), that can be passed to the callback/promise — this would ensure consistency in the methodologies of handling errors, as one can use the same syntax for filtering all failures — rather than having to use both .catch for operational errors and then couple that with try...catch for programmer errors.

**Comment by petkaantonov** [comment]
A completely separate module can be blacklisted so that  concerns such as promises ruining existing post-mortem workflows with callbacks can be addressed by simply blacklisting the separate module as people with such concerns already have to do for 3rd party modules.

**Comment by omsmith** [comment]
> +1 for require("fs/promise") vs require("fs").promised blocking unflagging but nor PR.

I'll echo that.

Don't think I'd want to see `.promised` end up as an actual API, but if it's felt it's a necessary stepping stone at this point - go for it.

**Comment by balupton** [comment]
Regarding the surface promise API, the only other suggestion that is not covered is just updating the existing API methods to use promises instead or with callbacks, so no new methods are introduced, just existing methods are updated. This has probably been discussed somewhere, if so would be good to have the conclusion summarised in the first/original post.

Proposals:

- `require('fs').readFile` returns promises only, no support for callback
- `require('fs').readFile` returns promise if no callback is provided, if callback is provided do not return promise and use callback
- `require('fs').readFile` returns promise always, if callback is provided attach it to the promise

**Comment by benjamingr** [comment]
> Regarding the surface promise API, the only other suggestion that is not covered is just updating the existing API methods to use promises instead or with callbacks, so no new methods are introduced

This was discussed already in the pull request. Namely https://github.com/nodejs/node/pull/5020#discussion_r51464838 

We're also not going to break "fs" or any of the existing modules.

**Comment by omsmith** [comment]
@chrisdickinson Apologies if I missed it in the PR thread (was hard to track all that). Would you (or someone else) be able to sum up:

* Is the reasoning of wanting to land with `.promised` and follow-up with a `fs/promise` is just to see _something_ land?
* Is it believed `.promised` will be more likely to get landed than a separate module?

**Comment by balupton** [comment]
@benjamingr cheers, I've updated the initial post with the counters so future people know they are ruled out and why

**Comment by balupton** [comment]
@omsmith

> Is the reasoning of wanting to land with .promised and follow-up with a fs/promise is just to see something land?

Seems so. I've updated the top post with the relevant links.

> Is it believed .promised will be more likely to get landed than a separate module?

I'm too confused about any difference in benefits between `require('fs').readFilePromise` and `require('fs').promised.readFile` - haven't spotted anything on it. More details on this would be nice.

**Comment by omsmith** [comment]
Benefits of having `fs/promise` or the `.promised` shortcut:
* nit: I don't have `Async` or `Promise` all over
* nit: The presented functions are _only_ `Promise` returning ones, which is what I actually wanted

Benfits of having __only__ `fs/promise`:
* Callback consumers do not have to load promisified functions into memory

**Comment by balupton** [comment]
Hrmm...

If `fs/promise` also supports callbacks (probably by attaching the callback to the promise if it exists), then one just ever needs to include it for both callback and promise support, never needing to include both `require('fs')` and `require('fs/promised')` if for some reason the codebase is large and they need both

> nit: The presented functions are only Promise returning ones, which is what I actually wanted

Another side of this coin is that considering the `process.send(message[, sendHandle][, callback])` return type issue as raised in https://github.com/nodejs/node/pull/5020#discussion_r51466441 - a promise version of `process.send` will break return types, so `require('process/promise').send` o which is now promised base will need different documentation, so `require('process').sendPromised` could provide a sense of false compatibility

**Comment by pesho** [comment]
> Benefits of having only `fs/promise`:

Also:
* Ability to blacklist promisified APIs altogether (mentioned by @petkaantonov above)

**Comment by balupton** [comment]
> Callback consumers do not have to load promisified functions into memory

Seems a strange argument as one only needs at least two modules in their entire app where one module uses the promise API and another module uses the callback API, for this to benefit to be negated, which as time goes by will become highly likely. Unless one of the modules (probably the promise API) is specifically blacklisted somehow, which would cause module restrictions which is perhaps undesirable — is such a blacklisting ever likely to occur?

**Comment by omsmith** [comment]
Perhaps, but I think it follows along with some comments I saw in the original PR:
* Don't want to affect the callback functions, performance wise (increased base memory usage could be seen as an effect)
* Node.js is used in many different environments, including embedded systems (we should try to respect the restrictions of low-memory environments)
* A lot of work goes into reducing resource usage and start-up times
* A lot of work goes into making sure core maintains a small surface area (while `fs/promise` would still increase of surface area of _core_, it at least doesn't increase the surface area of `fs`)

My own thought:
* Bundlers such as Browserify provide shims for core modules like `fs`. Adding the `Promise` returning methods inline will bloat those bundles as well (although minimally, thanks to the way it's been approached by chrisdickinson)

**Comment by vdeturckheim** [comment]
Could it be possible to choose an alternative promise library ? In mongoose, it is possible to pass a constructor that will be used by the package.

Default would be to standard promises, but the possibility to do something like `methodReturningPromise.as(MyPromiseConstructor).then...` ?

Or better (with all alternatives in discussion):
* `require('fs').promised([MyPromiseConstructor])`
* `require('fs/promise').as(MyPromiseConstructor).readFile` where the `as` method would be optional

**Comment by benjamingr** [comment]
@vdeturckheim that was discussed and decided against (I think). Please see the original PR's comments.

**Comment by benjamingr** [comment]
@vdeturckheim that was discussed and decided against (I think). Please see the original PR's comments.

**Comment by kriskowal** [comment]
I am in favor of having separate require("promised/*") modules so that the promise oriented porcelain does not get tangled with the underlying fs module. The same pattern may one day be called upon to provide a compatibility layer for web streams, e.g., require("web/fs"), which might be both promise oriented and web streams oriented. Regardless, we need to avoid entangling these concerns with the underlying callback and event-emitter oriented design.

**Comment by vdeturckheim** [comment]
@benjamingr sorry my bad then.

**Comment by mikeal** [comment]
Keep in mind that using a top level namespace will break anyone using a module named that in the ecosystem, in this case https://www.npmjs.com/package/promised which doesn't seem to be in heavy use. We'd want to talk with the publisher about pulling it and npm about blocking it from being accessible in the future if we go this route.

**Comment by chrisdickinson** [comment]
@mikeal I've emailed the owner to see if they'd be willing to transfer the package.

**Comment by omsmith** [comment]
Just to note, that would only be necessary for `require('promised/foo'), as opposed to `require('foo/promised')` which I think has generally been the discussion point.

**Comment by chrisdickinson** [comment]
@omsmith Indeed, but given some thought I believe nesting under `promised` is going to be (technically) a bit more tenable than nesting under each module.

**Comment by benjamingr** [comment]
Update: The author of "promised" has transferred ownership to the Node foundation (for @chrisdickinson )

**Comment by chrisdickinson** [comment]
I've updated the +1/-1's for the module approach to reflect the concerns with the approach — in short, it will take a lot of intense discussion, and the concrete version of it will spur a lot of debate. Out of the interest of making this process sustainable for myself, the plan of record is to work on getting 5020 merged without a solution for modules, then work to get modules in behind the flag before unflagging in a separate PR. To do otherwise risks the work in nodejs/node#5020 and that we've done here overall, and I'm not comfortable doing that.

While I appreciate the summary of potential plans and their merits/demerits, I'd like to propose we make it reflect the following plan of action:

* Behind the `--enable-promises` flag, Introduce `*Promise` methods for most single-occurrence async operations in the Node API, excluding streams and stream-pair generating methods. This is nodejs/node#5020.
  * Once we have established that there are no _blocking_ concerns for merging behind a flag, seek buy-in from the Node collaborator group on merging.
* Once the PR is merged, I will immediately open a PR introducing a `require('promise/*')` set of modules behind the flag.
  * We will initially consider this PR "blocking unflagging".
* Once we've brought all issues marked "blocking unflagging" to a conclusion, then we will lift the flag. Official support of the API will begin one version later.

**Comment by petkaantonov** [comment]
@chrisdickinson I am not sure if you saw it in some other thread, but I suggested separate module because such a module can be blacklisted, just like I presume 3rd party promise modules are currently blacklisted by post-mortem users.

**Comment by chrisdickinson** [comment]
@petkaantonov Yep — I saw — the problem is that since there are promisified methods on class instances, disallowing access to promise methods via `require` won't be enough to blacklist them.

**Comment by omsmith** [comment]
@chrisdickinson which methods on which class instances?

**Comment by petkaantonov** [comment]
what about "promised/net" that returns "net" while adding promise methods on the first require?

**Comment by omsmith** [comment]
> -1 unclear on whether or not to re-export non-promise-returning methods (fs.createReadStream)

* What was the plan to export onto `.promised`?
* If we look to [mz/fs](https://www.npmjs.com/package/mz) a popular community effort, the non-promise-returning methods are excluded
  * May not be "bluebird-popular", but was the most popular "promisified fs" package I could find, outside of q-io, which is a little different
  * Falls in line with what I would expect
    * Would admittedly be a minor annoyance to have to import `createReadStream` seperately

**Comment by chrisdickinson** [comment]
@petkaantonov As I understand it, that would mean either we'd have to copy/re-run the `net` code to get our own copy of `Socket`, or that we'd mutate the non-promise-variant of `Socket`, which would be visible to other users.

**Comment by petkaantonov** [comment]
yes, thats what promisifyAll does. Having those methods only when a blacklistable module is required rather than all the time is surely better?

**Comment by chrisdickinson** [comment]
@petkaantonov To clarify, you're suggesting we mutate `require('net').Socket.prototype` when a user runs `require('promised/net')`? Or that we re-run the `net` code?

**Comment by petkaantonov** [comment]
Oh sorry I thought you were familiar with the api. It mutates the prototype (non destructively).
